general/forall.hpp

```diff
@@ -42,30 +42,30 @@ const int MAX_Q1D = 14;
 // The MFEM_FORALL wrapper
 #define MFEM_FORALL(i,N,...)                             \
    ForallWrap<1>(true,N,                                 \
-                 [=] MFEM_DEVICE (int i) {__VA_ARGS__},  \
-                 [&] MFEM_LAMBDA (int i) {__VA_ARGS__})
+                 [=] MFEM_DEVICE (int i) mutable {__VA_ARGS__},  \
```

Comment:
Why do the lambdas need to be mutable? Is it because of the non-const access to the Scalar data? Why was it not a problem before?

Reply:
Why is this necessary now?
We really need to make sure that this does not impact the performances before merging.

Reply:
lambda capture is `const` by default, so for instance:
```c++
auto X = Reshape(...);
```
then `X` is captured by the lambda as `const auto X`, independently of the `auto` type. In order to have lambda capturing variables with the same `const` value, one need to use `mutable` lambda capture, then we have:
```c++
X x; // <-- will be mutable in the lambda
const Y y; // <-- won't be mutable in the lambda
```
Currently, we're modifying `const` objects, which is not very clean IMHO.

Reply:
OK thanks!